core: introduce crossfile arrays (xfa) v3

kaspar030 opened the pull request:
# Contribution description

This PR is the third iteration of Cross File Arrays (xfa), which allow typed and sorted arrays to be spread over multiple .c files.

This is basically a rebased version of #9105.

## What can this be used for?

This could be useful to organize auto_init functions.

Currently, they're centrally managed in a single auto_init. file (well, multiple if auto_init_saul, _netif etc. should be counted).

With XFA's it is possible to add an initialization function to a global "init_functions" array, __within the module's .c file__:

     XFA_ADD_PTR(init_functions, 1234, ztimer, &ztimer_init)

`auto_init.c` could be reduced to this:

```
XFA_CONST_USE(init_function_t, init_functions);

void auto_init() {
    unsigned n = XFA_LEN(init_function_t, init_functions);
    for (unsigned i = 0; i < n; i++) {
		init_functions[i]();
    }
}
```

Compared to libc_init_array, XFA's support priority values. All entries get sorted by priority value, at link time.
In the above example, the ztimer init function was added with priority 100. Anything requiring ztimer would need to use a priority value that is higher than that.
The priority space is basically unlimited, the numbers get turned into symbol names, which the linker sorts. (note to self: see if that sort is numerical, or maybe 10 sorts before 2).

This is also not limited to functions. Other candidates for benefiting from non-central arrays are:
- shell commands
- saul drivers
- netif drivers
- threads that should started at boot time

# Downsides

Apart from being a bit macro hacky, this currently requires disabling `-Warray-bounds` for files using XFA's (that include xfa.h).
Also, this needs linker script support. A generic `xfa.ld` is used unless overridden. Some architectures need slightly different handling due to different section names or linker symbols.

### Testing procedure

As is, XFA's are not used anywhere other than in the test.
Ensure that test works, on all platforms.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

### Issues/PRs references

Taken over #9105.
Initially introduced in #7523.
Not needing ugly hacks since #14754.
 
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

maribu commented:
> Apart from being a bit macro hacky, this currently requires disabling -Warray-bounds for files using XFA's

Would it be possible to add a `static inline` helper function for accessing the arrays? If so, maybe some `#pragma GCC diagnostics` dance around that helper function would do?

kaspar030 commented:
> Would it be possible to add a `static inline` helper function for accessing the arrays? If so, maybe some `#pragma GCC diagnostics` dance around that helper function would do?

it probably would. also, this could make it possible to not force avr to do this in completely RAM. It would not make it more ergonomic, I think. will need to try.

kaspar030 commented:
I don't manage to adapt the xfa.ld to the internal msp430 linker script. :( some help would be appreciated.

I see some options:

- someone with more ld-script-fu figures out how to do it
- we extract the msp430 ld scripts and add them to our repo, then add the xfa support there (there are like 5 different ones)
- we make xfa a feature for now, excluding msp430, making it much less usable

maribu commented:
The ld-script-force is strong with @benpicco. Maybe he can help?

benpicco commented:
I don't have the hardware to test, but when I drop d50d0b6de7d381a95785a9707cd9bac6fa9dca1b and apply this patch

```patch
diff --git a/makefiles/arch/msp430.inc.mk b/makefiles/arch/msp430.inc.mk
index 011e2778ea..2210919568 100644
--- a/makefiles/arch/msp430.inc.mk
+++ b/makefiles/arch/msp430.inc.mk
@@ -15,6 +15,7 @@ ASFLAGS += $(CFLAGS_CPU) --defsym $(CPU_MODEL)=1 $(CFLAGS_DBG)
 
 LINKFLAGS += $(CFLAGS_CPU) $(CFLAGS_DBG) $(CFLAGS_OPT)
 LINKFLAGS += -Wl,--gc-sections -Wl,-L$(MSP430_SUPPORT_FILES)/include
+LINKFLAGS += -T $(MSP430_SUPPORT_FILES)/include/$(CPU_MODEL).ld
 
 OPTIONAL_CFLAGS_BLACKLIST += -fdiagnostics-color
 OPTIONAL_CFLAGS_BLACKLIST += -Wformat-overflow
```

it will at least compile for msp430

jia200x commented:
> I don't have the hardware to test, but when I drop d50d0b6 and apply this patch

I have a z1 lying around. I will try it

kaspar030 commented:
Seems like the riotdocker llvm is too old for the static test:

```
llvm-objdump: Unknown command line argument '--syms'.  Try: '/usr/bin/llvm-objdump -help'
llvm-objdump: Did you mean '-dsym'?
llvm-objdump: Unknown command line argument '--syms'.  Try: '/usr/bin/llvm-objdump -help'
llvm-objdump: Did you mean '-dsym'?
Error: Static check of XFA linked const array failed, verify linker flags and try again
```

The test succeeds with llvm 11, but not with the one in our build container.

fjmolinas commented:
@kaspar030 I can try again on a `z1`, le me know when the offline comment you mentioned are addressed and a new test is worth it.

kaspar030 commented:
> @kaspar030 I can try again on a `z1`, le me know when the offline comment you mentioned are addressed and a new test is worth it.

please try, it should work now!

fjmolinas commented:
Testing results on z1:

- `BOARD=z1 make -C tests/xfa/ flash test`

```
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-480-gb6725-pr-15002)
Cross file array test
xfatest[2]:
[0] = 1, "xfatest1"
[1] = 2, "xfatest2"
xfatest_const[2]:
[0] = 123, "xfatest_const1"
[1] = 45, "xfatest_const2"
```

- `BOARD=z1 make -C tests/unittests/ tests-core clean flash test`

```
READY
s
START
..................................................................................................
core_xfa_tests.test_xfa_const_data (tests/unittests/tests-core/tests-core-xfa.c 75) exp 2 was 0
..
run 100 failures 1
```

An issue in the unit-test pops out.

kaspar030 commented:
I think I found the cause. Seems like the newer gcc as used for msb430 figured out that a const rodata array with zero size is just that. I added a commit marking const XFA's "volatile", this makes the unittests pass for z1.

@fjmolinas can you take another look?

(I also rebased this before testing)

fjmolinas commented:
All passing now!:

`BOARD=z1 make -C tests/unittests/ tests-core clean flash test`

```
START
....................................................................................................
OK (100 tests)
```

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,161 @@
+ * index out of bounds" warning.  So until a solution for that is found, we
+ * need to disable array bounds checks for files using XFAs.
+ */
+_Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
```
Can we use `#pragma GCC diagnostic push` so its contained to this file?

kaspar030 replied:
no, because the users (files that include) do the accesses. :(

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,161 @@
+ */
+#define XFA_INIT_CONST(type, name) \
+    _Pragma("GCC diagnostic push") \
+    _Pragma("GCC diagnostic ignored \"-Wpedantic\"") \
```
What does the compiler complain about here?

kaspar030 replied:
IIRC, zero sized arrays

kaspar030 replied:
I double checked and added comments

fjmolinas commented on tests/xfa/Makefile:
```diff
@@ -0,0 +1,13 @@
+# name of your application
```
```suggestion
```

fjmolinas commented on tests/xfa/tests/01-run.py:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2018 Gaëtan Harter <gaetan.harter@fu-berlin.de>
```
wrong copyright?

kaspar030 replied:
no, Gaëtan wrote the test :)

fjmolinas commented:
> The ld-script-force is strong with @benpicco. Maybe he can help?

Since the ld-script-force is strong with @benpicco might giving this one a look?

kaspar030 commented:
note to self: https://github.com/RIOT-OS/RIOT/pull/15002#issuecomment-721630966 is still an issue

miri64 commented:
> note to self: [#15002 (comment)](https://github.com/RIOT-OS/RIOT/pull/15002#issuecomment-721630966) is still an issue

You could blacklist the LLVM toolchain when `xfa` is used.

kaspar030 commented:
> You could blacklist the LLVM toolchain when `xfa` is used.

hmja, but that reduces usability quite a bit.
Maybe we can just skip the test when llvm-objdump is too old.

benpicco commented:
Would https://github.com/RIOT-OS/riotdocker/pull/115 provide the solution?

kaspar030 commented:
> Would [RIOT-OS/riotdocker#115](https://github.com/RIOT-OS/riotdocker/pull/115) provide the solution?

I think so.

kaspar030 commented:
> note to self: [#15002 (comment)](https://github.com/RIOT-OS/RIOT/pull/15002#issuecomment-721630966) is still an issue

seems like "--syms" is an alias for "-t", which is available in both binutils and llvm objdump. Only the alias is missing in older llvm versions.

fjmolinas commented:
> > Would it be possible to add a `static inline` helper function for accessing the arrays? If so, maybe some `#pragma GCC diagnostics` dance around that helper function would do?
> 
> it probably would. also, this could make it possible to not force avr to do this in completely RAM. It would not make it more ergonomic, I think. will need to try.

@kaspar030 did you end up trying this out?

fjmolinas commented:
please squash @kaspar030

kaspar030 commented:
> @kaspar030 did you end up trying this out?

Nope! I think we should maybe find a use for this first?

kaspar030 commented:
> please squash @kaspar030

squashed & rebased.

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,153 @@
+ * This macro, in combination with an entry in the linker scripts, allows the
+ * definition of constant arrays to be spread over multiple C compilation
+ * units. These arrays are called "cross-file arrays" or short xfa.
+ *
+ *
+ *
```
Some extra lines here

fjmolinas reviewed (commented):
@kaspar030 some static checks are failing, mind re-triggering the ci after addressing?

fjmolinas commented:
Also af5db89255480bb5f8759e5a18d5a1e7845035eb should be squashed out IMO

fjmolinas commented:
Is there value in keeping 94e1b375a3dae38a82bb68ac4359e5375538dd49 unsquashed?

fjmolinas commented:
> Is there value in keeping [94e1b37](https://github.com/RIOT-OS/RIOT/commit/94e1b375a3dae38a82bb68ac4359e5375538dd49) unsquashed?

Same for 7383e1968c25c539d13108761a8e6a6856cfe4ed. I let you judge if its pertinent to keep those in history.

kaspar030 commented:
> Also [af5db89](https://github.com/RIOT-OS/RIOT/commit/af5db89255480bb5f8759e5a18d5a1e7845035eb) should be squashed out IMO

that one I squashed out. the others don't apply cleanly to e.g., the first commit. Do you mind?

kaspar030 commented:
For the static tests I need some help, I don't understand the doxygen errors. To me the members look documented?

fjmolinas commented:
I also see static tests are  mentioning:

```
Command output:

	file core/include/xfa.h needs to be uncrustified.
	Please run 'dist/tools/uncrustify/uncrustify.sh'
```

fjmolinas commented on tests/xfa/xfatest.h:
```diff
@@ -0,0 +1,29 @@
+ * directory for more details.
+ */
+
+#ifndef DOXYGEN /* just a test header, please ignore */
```
I think you need to move this after `XFATEST_H`

fjmolinas replied:
And after `#ifdef __cplusplus`

kaspar030 commented:
> I also see static tests are mentioning:
> 
> ```
> 
> ```

this is a warning only. The files are uncrustified with my local uncrustify version...

kaspar030 commented:
These I don't understand:

```
 Running "./dist/tools/doccheck/check.sh" x
Command output:

	ERROR: Doxygen generates the following warnings:
	core/include/xfa.h:38: warning: Member _XFA(name, prio) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:46: warning: Member _XFA_CONST(name, prio) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:63: warning: Member XFA_INIT_CONST(type, name) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:85: warning: Member XFA_INIT(type, name) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:104: warning: Member XFA_USE_CONST(type, name) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:119: warning: Member XFA_USE(type, name) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:135: warning: Member XFA(xfa_name, prio) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:149: warning: Member XFA_CONST(xfa_name, prio) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:161: warning: Member XFA_ADD_PTR(xfa_name, prio, name, entry) (macro definition) of file xfa.h is not documented.
	core/include/xfa.h:169: warning: Member XFA_LEN(type, name) (macro definition) of file xfa.h is not documented.
```

fjmolinas commented:
> These I don't understand:
> 
> ```
>  Running "./dist/tools/doccheck/check.sh" x
> Command output:
> 
> 	ERROR: Doxygen generates the following warnings:
> 	core/include/xfa.h:38: warning: Member _XFA(name, prio) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:46: warning: Member _XFA_CONST(name, prio) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:63: warning: Member XFA_INIT_CONST(type, name) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:85: warning: Member XFA_INIT(type, name) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:104: warning: Member XFA_USE_CONST(type, name) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:119: warning: Member XFA_USE(type, name) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:135: warning: Member XFA(xfa_name, prio) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:149: warning: Member XFA_CONST(xfa_name, prio) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:161: warning: Member XFA_ADD_PTR(xfa_name, prio, name, entry) (macro definition) of file xfa.h is not documented.
> 	core/include/xfa.h:169: warning: Member XFA_LEN(type, name) (macro definition) of file xfa.h is not documented.
> ```

me neither...

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,181 @@
+}
+#endif
+
+/** @} */
+#endif /* XFA_H */
```
```suggestion
#endif /* XFA_H */
/** @} */
```

bergzand commented:
Guarding the `_Pragma` with an `ifndef DOXYGEN` fixed the rendering for me. Let's see if that also silences the errors

bergzand commented:
It at least improves the situation

bergzand commented:
And that should fix the remaining doxygen issues

maribu commented on tests/unittests/tests-core/tests-core-xfa.c:
```diff
@@ -0,0 +1,141 @@
+            case 12345:
+                /* tests-core-xfa-data1.c */
+                TEST_ASSERT_EQUAL_STRING("xfatest1", xfatest[k].text);
+                ++found;
```
```suggestion
                TEST_ASSERT(!(found & BIT0));
                found |= BIT0;
```

Maybe it is better to use a bit mask here rather than a counter. If due to an error the value `12345` gets added two times but `0xbeef` is missing, the test would still pass.

maribu commented on tests/unittests/tests-core/tests-core-xfa.c:
```diff
@@ -0,0 +1,141 @@
+            case 0xbeef:
+                /* tests-core-xfa-data2.c */
+                TEST_ASSERT_EQUAL_STRING("another test string", xfatest[k].text);
+                ++found;
```
```suggestion
                TEST_ASSERT(!(found & BIT1));
                found |= BIT1;
```

maribu commented on tests/unittests/tests-core/tests-core-xfa.c:
```diff
@@ -0,0 +1,141 @@
+                break;
+        }
+    }
+    TEST_ASSERT_EQUAL_INT(n, found);
```
```suggestion
    TEST_ASSERT_EQUAL_INT((1U << (n + 1)) - 1, found);
```

kaspar030 commented:
- rebased
- added a note clarifying this is experimental

maribu reviewed (commented):
Looks good to me. Maybe the unit test could be spiced up a bit as suggested above, but I'm not insisting.

maribu commented:
If you take the suggestions to the unit test, please squash right away.

kaspar030 commented:
> If you take the suggestions to the unit test, please squash right away.

done

maribu commented on tests/unittests/tests-core/tests-core-xfa.c:
```diff
@@ -0,0 +1,143 @@
+ */
+
+#include <stdint.h>
+#include "xfa.h"
```
```suggestion
#include "bitarithm.h"
#include "xfa.h"
```

Please squash right away.

maribu commented on tests/unittests/tests-core/tests-core-xfa.c:
```diff
@@ -0,0 +1,144 @@
+                break;
+        }
+    }
+    TEST_ASSERT_EQUAL_INT((1U << (n + 1)) - 1, found);
```
```suggestion
    TEST_ASSERT_EQUAL_INT((1U << n) - 1, found);
```

Sorry, my bad. I should better leave calculations to the machines :-D

maribu commented:
Same issue for the Arduino Mega2560 with `BUILD_IN_DOCKER=1`. Blacklist AVR and get it in?

kaspar030 commented:
> Same issue for the Arduino Mega2560 with `BUILD_IN_DOCKER=1`. Blacklist AVR and get it in?

hm, seems like a deeper issue. this branch cannot flash other Arduino Mega2560 either. I'll dig in.

kaspar030 commented:
I solved the AVR error. Seems like a simple "-T" was missing for the xfa.ld include. I tested hello-world and the xfa test on atrduino-mega2560.

kaspar030 commented:
may I squash this as long as it is way up the queue?

fjmolinas commented:
> may I squash this as long as it is way up the queue?

Go ahead.

fjmolinas commented:
Re-tested all passing, lets wait for the Murdock fix!

fjmolinas commented:
All green!!

fjmolinas commented on tests/xfa/tests/01-run.py:
```diff
@@ -0,0 +1,26 @@
+    child.expect_exact('[1] = 45, "xfatest_const2"')
+
+
+if __name__ == "__main__":
+    sys.path.append(os.path.join(os.environ['RIOTTOOLS'], 'testrunner'))
+    from testrunner import run
+    exit(run(testfunc))
```
This is some old styling fro how it is done now..

fjmolinas commented on tests/xfa/xfatest.h:
```diff
@@ -0,0 +1,29 @@
+extern "C" {
+#endif
+
+#ifndef DOXYGEN /* just a test header, please ignore */
```
FUnny why doxygen complained about ths and not the unittest header.

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,186 @@
+ * @param[in]   name        symbol name
+ * @param[in]   entry       pointer variable to add to xfa
+ */
+#define XFA_ADD_PTR(xfa_name, prio, name, entry) \
```
There is an uncrustify comment if you want to address it.

kaspar030 replied:
ran uncrustify over it

fjmolinas commented on core/include/xfa.h:
```diff
@@ -0,0 +1,186 @@
+ * definition of constant arrays to be spread over multiple C compilation
+ * units. These arrays are called "cross-file arrays" or short xfa.
+ *
+ * @note This API is considered experimental and will probably change without
+ *       notice!
```
```suggestion
 * @experimental This API is considered experimental and will probably change without
 *       notice!
```

fjmolinas commented:
I noticed a final round of nitpicks.

fjmolinas commented:
@kaspar030 Murdock is idle, good time to address those last couple of nitpicks ;)

kaspar030 commented:
> @kaspar030 Murdock is idle, good time to address those last couple of nitpicks ;)

done & squashed

fjmolinas reviewed (approved):
ACK!

kaspar030 commented:
Thanks a lot everyone involved and especially @fjmolinas for reviewing, and for keeping me looking at this!

miri64 commented:
:tada:
